balancer/rls/balancer_test.go

```diff
@@ -650,6 +650,163 @@ func (s) TestConfigUpdate_DataCacheSizeDecrease(t *testing.T) {
+		}
+		if req.KeyMap["k2"] == "v2" {
+			return &rlstest.RouteLookupResponse{Resp: &rlspb.RouteLookupResponse{Targets: []string{backendAddress2}}}
+		}
+		return &rlstest.RouteLookupResponse{Err: errors.New("no keys in request metadata")}
+	})
+
+	// Register a manual resolver and push the RLS service config through it.
+	r := manual.NewBuilderWithScheme("rls-e2e")
+	scJSON := fmt.Sprintf(`
+{
+  "loadBalancingConfig": [
+    {
+      "%s": {
+		"routeLookupConfig": {
+			"grpcKeybuilders": [{
+				"names": [{"service": "grpc.testing.TestService"}],
+				"headers": [
+				   {
+					 "key": "k1",
+					 "names": [
+						   "n1"
+					 ]
+				   },
+				   {
+					 "key": "k2",
+					 "names": [
+						   "n2"
+					 ]
+				   }
+				]
+			}],
+			"lookupService": "%s",
+			"cacheSizeBytes": 1000
+		},
+		"childPolicy": [{"%s": {}}],
+		"childPolicyConfigTargetFieldName": "Backend"
+      }
+    }
+  ]
+}`, topLevelBalancerName, rlsServer.Address, childPolicyName)
+	sc := internal.ParseServiceConfig.(func(string) *serviceconfig.ParseResult)(scJSON)
+	r.InitialState(resolver.State{ServiceConfig: sc})
+
+	cc, err := grpc.Dial(r.Scheme()+":///", grpc.WithResolvers(r), grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		t.Fatalf("grpc.Dial() failed: %v", err)
+	}
+	defer cc.Close()
+
+	<-clientConnUpdateDone
+
+	// Make an RPC and ensure it gets routed to the first backend.
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	ctxOutgoing := metadata.AppendToOutgoingContext(ctx, "n1", "v1")
+	makeTestRPCAndExpectItToReachBackend(ctxOutgoing, t, cc, backendCh1)
+
+	// Make sure an RLS request is sent out.
+	verifyRLSRequest(t, rlsReqCh, true)
+
+	testutils.AwaitState(ctx, t, cc, connectivity.Ready)
+
+	// Cache the state changes seen up to this point.
+	states0 := ccWrapper.getStates()
+
+	// Make another RPC with a different set of headers. This will force the LB
+	// policy to send out a new RLS request, resulting in a new data cache
+	// entry.
+	ctxOutgoing = metadata.AppendToOutgoingContext(ctx, "n2", "v2")
+	makeTestRPCAndExpectItToReachBackend(ctxOutgoing, t, cc, backendCh2)
+
+	// Make sure an RLS request is sent out.
+	verifyRLSRequest(t, rlsReqCh, true)
+
+	// We currently have two cache entries. Setting the size to 1, will cause
+	// the entry corresponding to backend1 to be evicted.
+	rlsConfig.RouteLookupConfig.CacheSizeBytes = 1
+
+	// Push the config update through the manual resolver.
+	scJSON1, err := rlsConfig.ServiceConfigJSON()
+	if err != nil {
+		t.Fatal(err)
+	}
+	sc1 := internal.ParseServiceConfig.(func(string) *serviceconfig.ParseResult)(scJSON1)
+	r.UpdateState(resolver.State{ServiceConfig: sc1})
+
+	<-clientConnUpdateDone
+
+	states1 := ccWrapper.getStates()
+	if len(states1) == len(states0)+1 {
+		t.Fatalf("more than one state update seen. before %v, after %v", states0, states1)
```

Comment:
nit: It may be desirable to continue the test so using t.Errorf may be a better choice here.

See https://google.github.io/styleguide/go/decisions#keep-going